folly/SocketPortability.h

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2015 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// While MSVC may not have Unix sockets, Cygwin does.
+#ifdef _MSC_VER
+# define HAVE_UNIX_SOCKETS 0
+#else
+# define HAVE_UNIX_SOCKETS 1
+#endif
+
+#ifdef _MSC_VER
+
+#ifndef __STDC__
+#define __STDC__ 1
+#include <io.h>
+#undef __STDC__
+#else
+#include <io.h>
```

Comment:
For the wrappers to work correctly, as I have to implement `close` in a way that also supports socket handles, I need to make sure the defines that remap them to `_*` versions are not used. Those defines are disabled by defining `__STDC__`, however, if I were to define that for everything, I would need a whole lot more wrappers, because that define would disable just about every standard C function if it's not prefixed with an underscore. This means that I have to define it for a select few files. If `__STDC__` is already defined however, I don't need to redefine it. Having the include in multiple places is shorter than wrapping the def/udef pair.